src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.CrossTarget;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    /// <summary>
+    /// Model for creating telemetry events when dependency tree is updated.
+    /// It maintains some light state for each Target Framework to keep track
+    /// of the status of Rule processing so we can heuristically determine
+    /// when the dependency tree is "ready"
+    /// </summary>
+    [Export(typeof(IDependencyTreeTelemetryService))]
+    [AppliesTo(ProjectCapability.DependenciesTree)]
+    internal class DependencyTreeTelemetryService : IDependencyTreeTelemetryService
+    {
+        private const string TelemetryEventName = "TreeUpdated";
+        private const string UnresolvedLabel = "Unresolved";
+        private const string ResolvedLabel = "Resolved";
+        private const string ProjectProperty = "Project";
+
+        /// <summary>
+        /// Maintain state for each target framework
+        /// </summary>
+        public class TelemetryState
+        {
+            public ConcurrentDictionary<string, bool> ObservedRuleChanges { get; } = new ConcurrentDictionary<string, bool>();
```

Comment:
>ConcurrentDictionary [](start = 89, length = 20)

Just checking: rules do have constant names usually, but do we really want to have case sensitive dictionary here?